compiler/rustc_codegen_ssa/src/back/write.rs

```diff
@@ -1037,7 +1037,8 @@ fn start_executing_work<B: ExtraBackendMethods>(
     }));
 
     let ol = if tcx.sess.opts.debugging_opts.no_codegen
-        || !tcx.sess.opts.output_types.should_codegen()
+        || (!tcx.sess.opts.output_types.should_codegen()
+            && !tcx.sess.opts.debugging_opts.metadata_link)
```

Comment:
This shouldn't be necessary. This is only relevant for the actual codegen, not for encoding metadata.